keylime/ima/ima.py

```diff
@@ -477,7 +498,54 @@ def read_allowlist(al_path=None, checksum="", gpg_sig_file=None, gpg_key_file=No
                 f"Checksum of allowlist does not match! Expected {checksum}, Calculated {calculated_checksum}"
             )
 
+    # Return artifacts as a JSON bundle, ready for sending
+    bundle = {
+        "ima_policy": base64.b64encode(alist_bytes).decode(),
+        "key": base64.b64encode(al_key).decode(),
+        "sig": base64.b64encode(al_sig).decode(),
+        "checksum": checksum,
+    }
+
+    return bundle
+
+
+class SignatureValidationError(Exception):
+    def __init__(self, message, code):
+        self.message = message
+        self.code = code
+        super().__init__(self.message)
+
+
+# Reads allowlist bundles sent to the verifier, validates them, and processes them.
+def unbundle_allowlist(allowlist_bundle, force_verification=False):
+    logger.info("allowlist bundle: %s", allowlist_bundle)
+    allowlist_raw = base64.b64decode(allowlist_bundle["ima_policy"])
```

Comment:
So the new API requires to always send a bundle with at least the `ima_policy` inside it. This needs then to be documented somewhere what is optional and what not.

Reply:
@THS-on Not necessarily - this code is only called if a valid bundle was found in the first place: https://github.com/keylime/keylime/pull/994/files#diff-67033196f7dec303b3ea55793c6c13fb4b03a72a5515239475a871c3d638dfe4R428-R430

If not, the verifier will now populate the allowlist entry in the database with a default allowlist: https://github.com/keylime/keylime/pull/994/files#diff-67033196f7dec303b3ea55793c6c13fb4b03a72a5515239475a871c3d638dfe4R436-R438

I'm not sure if this is behavior we want, but I believe tests break if we don't put anything in the database for `allowlist`. Open to suggestions.

Reply:
What I meant is that I cannot post the old allowlist format to the verifier anymore I need to wrap it in a bundle when I want to use the IMA feature, which is fine.

> I'm not sure if this is behavior we want, but I believe tests break if we don't put anything in the database for allowlist. Open to suggestions.

This is fine for now.

Reply:
@THS-on My latest push fixes this - it will now maintain backwards compatibility with the old allowlist format without a bundle.